### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'feed' (possibly race condition on async disconnect)

### ISSUE THREAD

**Issue by bentheiii** [redis/redis-py#2349]
**Version**: 4.3.4
**Platform**: python 3.8.13 on azure

**Description**: unfortunately I have not been able to replicate this error locally, I'm guessing it's a bizarre race condition (possibly an async resurgence of #732)

```
AttributeError: 'NoneType' object has no attribute 'feed'
  File "***", line **
    await pipeline.execute()
  File "redis/asyncio/client.py", line 1352, in execute
    return await conn.retry.call_with_retry(
  File "redis/asyncio/retry.py", line 59, in call_with_retry
    return await do()
  File "redis/asyncio/client.py", line 1266, in _execute_pipeline
    await self.parse_response(connection, args[0], **options)
  File "redis/asyncio/client.py", line 1291, in parse_response
    result = await super().parse_response(connection, command_name, **options)
  File "redis/asyncio/client.py", line 505, in parse_response
    response = await connection.read_response()
  File "redis/asyncio/connection.py", line 931, in read_response
    response = await self._parser.read_response(
  File "redis/asyncio/connection.py", line 526, in read_response
    await self.read_from_socket()
  File "redis/asyncio/connection.py", line 491, in read_from_socket
    self._reader.feed(buffer)
```

**Comment by emtrane** [comment]
I'm able to reproduce the same very consistently with a script like this.
```
import asyncio
import wrapt
from redis.asyncio.cluster import RedisCluster

async def main():
  host = "localhost"
  port = 6379
  MAX_HASH_SLOTS = 16384

  def _execute_command_wrapper():
      async def _wrapped(f, instance, args, kwargs):
          if len(args) == 1:
              if args[0] == "INFO":
                  return {"cluster_enabled": True}
              elif args[0] == "CLUSTER SLOTS":
                  return [[0, MAX_HASH_SLOTS - 1, [host, port]]]

          return await f(*args, **kwargs)

      return _wrapped

  wrapt.wrap_function_wrapper(
      "redis.asyncio.cluster", "ClusterNode.execute_command", _execute_command_wrapper()
  )

  rc = RedisCluster(
    host=host, 
    port=port
  )

  async def t():
    for i in range(1000):
        async with rc.pipeline() as p:
            p.sadd('foo', 'bar')
            p.expire('foo', 3)
            p.incr('bar')
            p.expire('bar', 3)
            results = await p.execute()
            print(results)

  coros = [t() for i in range(1000)]
  await asyncio.gather(*coros)

asyncio.run(main())
```

**Comment by utkarshgupta137** [comment]
Can you guys report back the results with the latest pre-release?

**Comment by utkarshgupta137** [comment]
I'm unable to reproduce the error using the above code on my local mac with `v4.4.0rc1`.

**Comment by utkarshgupta137** [comment]
@emtrane Can you try creating a PR with the above code as a failing test?

**Comment by kristjanvalur** [comment]
Interesting, I'll try to reproduce and figure out what's going on.

**Comment by kristjanvalur** [comment]
What is happening is that someone (a task) disconnects the connection using `disconnect()`.  this starts by clearing the parser state before actually closing the socket. 
Meanwhile, a "read" on the socket finishes on a different task.  and it finds that the parser has been cleared.

Clearly thee intention here is that one Task can close a connection even if the other is busy reading, and what should
happen is that the parser, upon finding itself cleared, should raise a disconnect.  

There are two ways to fix this:
 - checking for the None and raising a Disconnection exception
 - calling `parser.on_disconnect()` in the finally clause of the `connection.disconnect()` thing.
 
 It should be possible to reproduce this problem using Mocks.

**Comment by juanamari94** [comment]
Hi @kristjanvalur , appreciate the insight! Would you like me to look into writing a unit test for this as @utkarshgupta137 mentioned? Or will you be taking this over?

**Comment by kristjanvalur** [comment]
I already have a unittest and a PR going on.  There are two approaches to fixing this, we can discuss in the PR when ready.

**Comment by dvora-h** [comment]
Fixed in #2557
